src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.CacheEntry.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion.Providers.ImportCompletion;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal static partial class ExtensionMethodImportCompletionHelper
+    {
+        private readonly struct CacheEntry
+        {
+            public Checksum Checksum { get; }
+            public string Language { get; }
+
+            /// <summary>
+            /// Mapping from the name of target type to extension method symbol infos.
+            /// </summary>
+            public readonly MultiDictionary<string, DeclaredSymbolInfo> SimpleExtensionMethodInfo { get; }
+
+            public readonly ImmutableArray<DeclaredSymbolInfo> ComplexExtensionMethodInfo { get; }
+
+            private CacheEntry(
+                Checksum checksum,
+                string language,
+                MultiDictionary<string, DeclaredSymbolInfo> simpleExtensionMethodInfo,
+                ImmutableArray<DeclaredSymbolInfo> complexExtensionMethodInfo)
+            {
+                Checksum = checksum;
+                Language = language;
+                SimpleExtensionMethodInfo = simpleExtensionMethodInfo;
+                ComplexExtensionMethodInfo = complexExtensionMethodInfo;
+            }
+
+            public class Builder : IDisposable
```

Comment:
in the future consider structs for your builders. it makes it really easy to just be practically alloc free (except for the reuslts of course).